### PR TITLE
Semidynamic vector improvements

### LINKED ISSUE #20551: Enable zero initialization for SemidynamicVector
## Reason
I was waiting on a libmesh update for this. I think what I need is in. I'll now add the option to zero initialize vector elements and and a `push_back` member function.

## Design
Use `if constexpr` and a bool template parameter to allow disabling zero init.

## Impact
Added capability.

### PR DESCRIPTION
**Pull request by dschwen**
Closes #20551

### CODE REVIEW

**Comment by recuero** [comment]
Nice!

File: framework/include/utils/MooseUtils.h [hunk 1]
```diff
@@ -1007,23 +1007,41 @@ class SemidynamicVector
   SemidynamicVector(std::size_t size) : Parent()
   {
     Parent::resize(size);
-    // TODO: uncomment this once https://github.com/libMesh/MetaPhysicL/pull/8
-    //       makes it all the way into MOOSE.
-    // for (const auto i : make_range(size))
-    //   _data[i] = {};
+    if constexpr (zero_init)
```
Comment (loganharbour) [inline framework/include/utils/MooseUtils.h:1010]: :smiling_face_with_three_hearts:

**Comment by dschwen** [comment]
Is the libmesh update in?

**Comment by loganharbour** [comment]
> Is the libmesh update in?

you're probably looking for #20580

**Comment by cticenhour** [comment]
@dschwen aiming to have it in tomorrow.

**Comment by mangerij** [comment]
What is a "semidynamic" vector?

**Comment by dschwen** [comment]
> What is a "semidynamic" vector?

A vector that has compile time allocated max elements of which a dynamic amount can be used. It looks like std::vector but has a fixed storage that is not heap allocated and thus faster. It is suitable for small vectors of which you know the upper bound in size.

File: framework/include/utils/MooseUtils.h [hunk 2]
```diff
@@ -997,7 +997,7 @@ buildRequiredMaterials(const Consumers & mat_consumers,
  * Note: due to an assertion bug in DynamicStdArrayWrapper we have to allocate one element more
```
Comment (recuero) [inline framework/include/utils/MooseUtils.h:997]: This is now available in MOOSE: can we modify the need to allocate an extra element?

File: unit/src/MooseUtilsTest.C [hunk 3]
```diff
@@ -459,4 +459,18 @@ TEST(MooseUtils, SemidynamicVector)
...
+
+  // test push back
+  test.push_back(100);
+  count = 0;
+  for (auto & i : ctest)
+    count += i;
+  EXPECT_EQ(count, 1 + 2 + 3 + 4 + 5 + 6 + 100);
+
+  // test emplace_back
+  test.emplace_back(200);
```
Comment (recuero) [inline unit/src/MooseUtilsTest.C:471]: Can the optional initialization be unit tested meaningfully?

Comment (dschwen) [reply unit/src/MooseUtilsTest.C:472]: Added a test

File: framework/include/utils/MooseUtils.h (hunk 2 above)
Comment (dschwen) [reply framework/include/utils/MooseUtils.h]: Removed

File: framework/include/utils/MooseUtils.h [hunk 4]
```diff
@@ -999,34 +999,32 @@ buildRequiredMaterials(const Consumers & mat_consumers,
...
+    if constexpr (value_init)
       for (const auto i : make_range(size))
-        _data[i] = {};
+        _data[i] = T{};
   }
 
   void resize(std::size_t new_size)
   {
-    const auto old_dynamic_n = Parent::size();
+    [[maybe_unused]] const auto old_dynamic_n = Parent::size();
```
Comment (recuero) [inline framework/include/utils/MooseUtils.h:1021]: Oh woow

**Review by recuero — APPROVED** [review]
👍 Good with me.

File: framework/include/utils/MooseUtils.h (hunk 4 above)
Comment (dschwen) [reply framework/include/utils/MooseUtils.h:1021]: I'm pretty certain the unused variable gets optimized away by the compiler if the value init template parameter is false. Unfortunately if constexpr has scope, so there is no elegant way to do this without code duplication, and I think the attribute is the right way to do this here.

**Pull request merged**